src/memory-analyzer/analyze_symbol.h

```diff
@@ -36,64 +39,141 @@ class gdb_apit;
   void add_assignment(const exprt &lhs, const exprt &value);
 
+  /// Iterate over \p array and fill its operands with the results of calling
+  ///   \ref get_expr_value on index expressions into \p expr.
+  /// \param expr: the expression to be analysed
+  /// \param array: array with zero-initialised elements
+  /// \param location: not used
+  /// return an array of the same type as \p expr filled with values from gdb
   exprt get_array_value(
     const exprt &expr,
     const exprt &array,
     const source_locationt &location);
 
+  /// Case analysis on the \ref typet of \p expr
+  /// 1) For integers, booleans, and enums: call \ref gdb_apit::get_value
+  ///    directly
+  /// 2) For chars: return the \p zero_expr
+  /// 3) For structs, arrays, and pointers: call their dedicated functions
+  /// \param expr: the expression to be analysed
+  /// \param zero_expr: zero of the same type as \p expr
+  /// \param location: not used
+  /// \return the value of the expression extracted from \ref gdb_apit
   exprt get_expr_value(
     const exprt &expr,
     const exprt &zero_expr,
     const source_locationt &location);
 
+  /// For each of the members of the struct: call \ref get_expr_value
+  /// \param expr: struct expression to be analysed
+  /// \param zero_expr: struct with zero-initialised members
+  /// \param location: not used
+  /// \return the value of the struct from \ref gdb_apit
   exprt get_struct_value(
     const exprt &expr,
     const exprt &zero_expr,
     const source_locationt &location);
 
+  /// Call \ref gdb_apit::get_memory on \p expr then split based on the
+  ///   points-to type being `char` type or not. These have dedicated functions.
+  /// \param expr: the input pointer expression
+  /// \param zero_expr: null-pointer (or its equivalent)
+  /// \param location: not used
+  /// \return symbol expression associated with the gdb-extracted memory
+  ///    location
   exprt get_pointer_value(
     const exprt &expr,
     const exprt &zero_expr,
     const source_locationt &location);
 
+  /// Similar to \ref get_char_pointer_value. Doesn't re-call
+  ///   \ref gdb_apit::get_memory, calls \ref get_expr_value on _dereferenced_
+  ///   \p expr (the result of which is assigned to a new symbol).
+  /// \param expr: the pointer expression to be analysed
+  /// \param memory_location: pointer value from \ref gdb_apit::get_memory
+  /// \param location: not used
+  /// \return symbol expression associated with \p memory_location
   exprt get_non_char_pointer_value(
     const exprt &expr,
     const std::string memory_location,
     const source_locationt &location);
 
+  /// If \p memory_location is found among \ref values then return the symbol
+  ///   expression associated with it.
+  /// Otherwise we add the appropriate \ref values mapping:
+  /// 1) call \ref gdb_apit::get_memory on the \p expr
+  ///    ??? Doesn't this have to be the same as \p memory_location
```

Comment:
`get_memory()` in fact returns a struct with four members, one of them being the memory address. In this case we need the field `string` which is an `optionalt` that contains the string the pointer points to if it points to a string.

Reply:
ok, thanks for the explanation.